Sphinx 6 upgrade post (#201)
readthedocs/blog: sphinx6-upgrade.rst

.. post:: January 4th, 2023
   :tags: sphinx, python
   :author: Ben
   :location: MLM

.. meta::
   :description lang=en:
      Sphinx 6 is out. We share our considerations about upgrading.

Sphinx 6 is out and has important breaking changes
==================================================

Sphinx 6 was released on December 29, 2022.
It contains a few major breaking changes that users should be aware of,
and some smaller new features as well.
Here are some of our considerations for getting ready for Sphinx 6:

- Sphinx 6 requires Python 3.8+. To enable this on Read the Docs, you will need to specify a Python version using the :ref:`readthedocs:config-file/v2:build.tools.python` setting.
- Sphinx 6 drops support of docutils 0.14, 0.15, 0.16 and 0.17. It now supports only docutils 0.18 or 0.19.
  We are finalizing Sphinx 6 support for `sphinx-rtd-theme`_, which currently only supports docutils 0.17.
  These updates are almost finished and are planned for version 1.2.0,
  which you can already try out by installing ``sphinx-rtd-theme==1.2.0rc2``.

> - Sphinx 6 is removing jQuery.

Author: ```suggestion
- Sphinx 6 removes jQuery.
```

- Sphinx 6 is removing jQuery.
  The JavaScript asset is easily added back using the new extension `sphinxcontrib-jquery`_.
  It is included automatically by `sphinx-rtd-theme`_, so if you are using our theme,
  you will also continue to have jQuery available in your documentation.
- Sphinx 6 removes all other bundled JavaScript libraries.
  If you depend on any of these,
  you may have to bundle them yourself or modify your implementation.

> For more information and full details you can read the `Sphinx changelog <https://www.sphinx-doc.org/en/master/changes.html#release-6-0-0-released-dec-29-2022>`_
> 
> Since most documentation projects rely on themes and extensions,
> we recommend that you do not upgrade to Sphinx 6 before you see announcements of Sphinx 6 support from these projects.

Reviewer: ```suggestion

For more information and full details you can read the `Sphinx changelog <https://www.sphinx-doc.org/en/master/changes.html#release-6-0-0-released-dec-29-2022>`_

Since most documentation projects rely on themes and extensions,
we recommend checking for Sphinx 6 support in those projects before upgrading.
```

.. post:: January 4th, 2022
   :tags: sphinx, python
   :author: Ben
   :location: MLM

.. meta::
   :description lang=en:

>       Sphinx 6 is out. Here are our thoughts about preparations that we
>       are making, which may affect projects looking to upgrade.

Author: ```suggestion
      Sphinx 6 is out. We share our considerations about upgrading.
```

  you will also continue to have jQuery available in your documentation.
- Sphinx 6 removes all other bundled JavaScript libraries.
  If you depend on any of these,
  you may have to bundle them yourself or modify your implementation.

For more information and full details you can read the `Sphinx changelog <https://www.sphinx-doc.org/en/master/changes.html#release-6-0-0-released-dec-29-2022>`_

Since most documentation projects rely on themes and extensions,
we recommend checking for Sphinx 6 support in those projects before upgrading.

.. _sphinxcontrib-jquery: https://pypi.org/project/sphinxcontrib.jquery/

Less maintenance work in the future
-----------------------------------

Sphinx 6 focuses on reducing the maintenance burden for the future.
This is done by simplifying the factor of supported docutils versions and Python versions.

Moving forwards,
the number of upgrade paths for documentation projects
are fewer and thus will lead to less choices for documentation owners and
maintainers of Sphinx themes and extensions.
This is a positive step forward for both documentation projects, themes and extensions.

> We recommend this upgrade in order to reduce complexities and future technical debt.

Author: ```suggestion
If your Sphinx extensions and theme support Sphinx 6, we recommend this upgrade in order to reduce complexities and future technical debt.
```

.. post:: January 4th, 2022

>    :tags: newsletter, python

Reviewer: This isn't a newsletter, though we do need to publish that as well soon :)

Author: Oh yes, that was in a blind spot, thanks!

Author: Fixed

   :author: Ben
   :location: MLM

.. meta::
   :description lang=en:
      Sphinx 6 is out. Here are our thoughts about preparations that we
      are making, which may affect projects looking to upgrade.

Sphinx 6 is out and has important breaking changes
==================================================

Sphinx 6 was released on December 29, 2022.
It contains a few major breaking changes that users should be aware of,
and some smaller new features as well.
Here are some of our considerations for getting ready for Sphinx 6:

> - Sphinx 6 requires Python 3.8+ and you will need to specify a Python version using the :ref:`readthedocs:config-file/v2:build.tools.python` setting.

Reviewer: ```suggestion
- Sphinx 6 requires Python 3.8+. To enable this on Read the Docs, you will need to specify a Python version using the :ref:`readthedocs:config-file/v2:build.tools.python` setting.
```

> - Sphinx 6 drops support of docutils 0.14, 0.15, 0.16 and 0.17.
> - Sphinx 6 requires docutils 0.18 or 0.19.

Reviewer: ```suggestion
- Sphinx 6 drops support of docutils 0.14, 0.15, 0.16 and 0.17. It now supports only docutils 0.18 or 0.19.
```

>   We are finalizing support for `sphinx-rtd-theme`_ which currently only supports docutils 0.17.
>   These updates are almost finished and are planned to arrive in sphinx-rtd-theme 1.2.0,
>   which you can already try out by installing ``sphinx-rtd-theme==1.2.0rc2``
>   which supports docutils 0.18.

Reviewer: ```suggestion
  We are finalizing Sphinx 6 support for `sphinx-rtd-theme`_, which currently only supports docutils 0.17.
  These updates are almost finished and are planned for version 1.2.0,
  which you can already try out by installing ``sphinx-rtd-theme==1.2.0rc2``.
```

- Sphinx 6 is removing jQuery.
  The JavaScript asset is easily added back using the new extension `sphinxcontrib-jquery`_.

>   We need it for `sphinx-rtd-theme`_ so if you are using the theme,
>   you will also continue to have jQuery available in your documentation.

Reviewer: ```suggestion
  It is included automatically by `sphinx-rtd-theme`_, so if you are using our theme,
  you will also continue to have jQuery available in your documentation.
```

- Sphinx 6 removes all other bundled JavaScript libraries.
  If you depend on any of these,
  you may have to bundle them yourself or modify your implementation.

> - See also the `Sphinx changelog <https://www.sphinx-doc.org/en/master/changes.html>`_

Reviewer: ```suggestion
For more information and full details you can read the `Sphinx changelog <https://www.sphinx-doc.org/en/master/changes.html#release-6-0-0-released-dec-29-2022>`_
```

> Since most documentation projects rely on themes and extensions, we recommend that you do not upgrade to Sphinx 6 before you see announcements of Sphinx 6 support in your themes and extensions.

Reviewer: ```suggestion
Since most documentation projects rely on themes and extensions,
we recommend that you do not upgrade to Sphinx 6 before you see announcements of Sphinx 6 support from these projects.
```

.. _sphinxcontrib-jquery: https://pypi.org/project/sphinxcontrib.jquery/

Less maintenance work in the future
-----------------------------------

Sphinx 6 focuses on reducing the maintenance burden for the future.
This is done by simplifying the factor of supported docutils versions and Python versions.

Moving forwards,
the number of upgrade paths for documentation projects
are fewer and thus will lead to less choices for documentation owners and
maintainers of Sphinx themes and extensions.
At Read the Docs, we support both our main theme, `sphinx-rtd-theme`_, and a number of extensions for Sphinx.
In light of that,

> we are direct beneficiaries of the reduced dependency matrix.

Reviewer: This is kind of a weird ending to a blog post. We should sign off with something directed at the user, not talking about ourselves in passive voice.

Author: I've tried with a different phrasing. I think it's great that Sphinx 6 is pushing for docutils 0.18+ and Python 3.8+ which I think we can fully back.

the number of upgrade paths for documentation projects
are fewer and thus will lead to less choices for documentation owners and
maintainers of Sphinx themes and extensions.
This is a positive step forward for both documentation projects, themes and extensions.

> We recommend this upgrade in order to reduce complexities and future technical debt.

Reviewer: This is a bit in conflict with the above line:

`we recommend that you do not upgrade to Sphinx 6 before you see announcements of Sphinx 6 support from these projects.`

We should probably clarify that language.

Author: ```suggestion
If your Sphinx extensions and theme support Sphinx 6, we recommend this upgrade in order to reduce complexities and future technical debt.
```
